File: apps/frontend/src/adapters/shared/private-assets.ts
import type {
  FundManager,
  NewFundManager,
  NewPrivateAsset,
  NewPrivateSnapshot,
  NewPrivateSubAsset,
  PrivateAsset,
  PrivateAssetCurrentTotals,
  PrivateAssetDetail,
  PrivateAssetHistoricalPoint,
  PrivateAssetListRow,
  PrivateSnapshot,
  PrivateSubAsset,
  UpdateFundManager,
  UpdatePrivateAsset,
  UpdatePrivateSnapshot,
  UpdatePrivateSubAsset,
} from "@/lib/types";

import { invoke } from "./platform";

export const listPrivateAssetRows = async (
  includeArchived = false,
): Promise<PrivateAssetListRow[]> => {
  return invoke<PrivateAssetListRow[]>("list_private_asset_rows", { includeArchived });
};

export const getPrivateAssetDetail = async (
  privateAssetId: string,
): Promise<PrivateAssetDetail | null> => {
  return invoke<PrivateAssetDetail | null>("get_private_asset_detail", { privateAssetId });
};

export const getPrivateAssetCurrentTotals = async (
  includeArchived = false,
): Promise<PrivateAssetCurrentTotals> => {
  return invoke<PrivateAssetCurrentTotals>("get_private_asset_current_totals", {
    includeArchived,
  });
};

export const getPrivateAssetHistoricalSeries = async (
  includeArchived = false,
): Promise<PrivateAssetHistoricalPoint[]> => {
  return invoke<PrivateAssetHistoricalPoint[]>("get_private_asset_historical_series", {
    includeArchived,
  });
};

export const listFundManagers = async (): Promise<FundManager[]> => {
  return invoke<FundManager[]>("list_fund_managers");
};

export const createFundManager = async (payload: NewFundManager): Promise<FundManager> => {
  return invoke<FundManager>("create_fund_manager", { payload });
};

export const updateFundManager = async (
  fundManagerId: string,
  payload: UpdateFundManager,
): Promise<FundManager> => {
  return invoke<FundManager>("update_fund_manager", { fundManagerId, payload });
};

export const createPrivateAsset = async (payload: NewPrivateAsset): Promise<PrivateAsset> => {
  return invoke<PrivateAsset>("create_private_asset", { payload });
};

export const updatePrivateAsset = async (
  privateAssetId: string,
  payload: UpdatePrivateAsset,
): Promise<PrivateAsset> => {
  return invoke<PrivateAsset>("update_private_asset", { privateAssetId, payload });
};

export const listPrivateSubAssets = async (privateAssetId: string): Promise<PrivateSubAsset[]> => {
  return invoke<PrivateSubAsset[]>("list_private_sub_assets", { privateAssetId });
};

export const createPrivateSubAsset = async (
  payload: NewPrivateSubAsset,
): Promise<PrivateSubAsset> => {
  return invoke<PrivateSubAsset>("create_private_sub_asset", { payload });
};

export const updatePrivateSubAsset = async (
  privateSubAssetId: string,
  payload: UpdatePrivateSubAsset,
): Promise<PrivateSubAsset> => {
  return invoke<PrivateSubAsset>("update_private_sub_asset", { privateSubAssetId, payload });
};

export const listPrivateSnapshots = async (privateAssetId: string): Promise<PrivateSnapshot[]> => {
  return invoke<PrivateSnapshot[]>("list_private_snapshots", { privateAssetId });
};

export const getLatestPrivateSnapshot = async (
  privateAssetId: string,
): Promise<PrivateSnapshot | null> => {
  return invoke<PrivateSnapshot | null>("get_latest_private_snapshot", { privateAssetId });
};

export const createPrivateSnapshot = async (
  payload: NewPrivateSnapshot,
): Promise<PrivateSnapshot> => {
  return invoke<PrivateSnapshot>("create_private_snapshot", { payload });
};

export const updatePrivateSnapshot = async (
  privateSnapshotId: string,
  payload: UpdatePrivateSnapshot,
): Promise<PrivateSnapshot> => {
  return invoke<PrivateSnapshot>("update_private_snapshot", { privateSnapshotId, payload });
};

File: apps/tauri/src/commands/private_assets.rs
use std::sync::Arc;

use crate::context::ServiceContext;
use tauri::State;
use wealthfolio_core::private_assets::{
    FundManager, NewFundManager, NewPrivateAsset, NewPrivateSnapshot, NewPrivateSubAsset,
    PrivateAsset, PrivateAssetCurrentTotals, PrivateAssetDetail, PrivateAssetHistoricalPoint,
    PrivateAssetListRow, PrivateSnapshot, PrivateSubAsset, UpdateFundManager, UpdatePrivateAsset,
    UpdatePrivateSnapshot, UpdatePrivateSubAsset,
};

fn ensure_private_assets_enabled(state: &ServiceContext) -> Result<(), String> {
    if state.get_private_assets_enabled() {
        Ok(())
    } else {
        Err("Private assets are disabled in settings.".to_string())
    }
}

#[tauri::command]
pub async fn list_private_asset_rows(
    include_archived: Option<bool>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<PrivateAssetListRow>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_asset_projection_service()
        .list_private_asset_rows(include_archived.unwrap_or(false))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_fund_managers(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<FundManager>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .list_fund_managers()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_private_asset_detail(
    private_asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<PrivateAssetDetail>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_asset_projection_service()
        .get_private_asset_detail(&private_asset_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_private_asset_current_totals(
    include_archived: Option<bool>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateAssetCurrentTotals, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_asset_projection_service()
        .get_private_asset_current_totals(include_archived.unwrap_or(false))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_private_asset_historical_series(
    include_archived: Option<bool>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<PrivateAssetHistoricalPoint>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_asset_projection_service()
        .get_private_asset_historical_series(include_archived.unwrap_or(false))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_fund_manager(
    payload: NewFundManager,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<FundManager, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .create_fund_manager(payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_fund_manager(
    fund_manager_id: String,
    payload: UpdateFundManager,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<FundManager, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .update_fund_manager(&fund_manager_id, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_private_asset(
    payload: NewPrivateAsset,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateAsset, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .create_private_asset(payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_private_asset(
    private_asset_id: String,
    payload: UpdatePrivateAsset,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateAsset, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .update_private_asset(&private_asset_id, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_private_sub_assets(
    private_asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<PrivateSubAsset>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .list_private_sub_assets(&private_asset_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_private_sub_asset(
    payload: NewPrivateSubAsset,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateSubAsset, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .create_private_sub_asset(payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_private_sub_asset(
    private_sub_asset_id: String,
    payload: UpdatePrivateSubAsset,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateSubAsset, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .update_private_sub_asset(&private_sub_asset_id, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_private_snapshots(
    private_asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<PrivateSnapshot>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .list_private_snapshots(&private_asset_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_latest_private_snapshot(
    private_asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<PrivateSnapshot>, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .get_latest_private_snapshot(&private_asset_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_private_snapshot(
    payload: NewPrivateSnapshot,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateSnapshot, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .create_private_snapshot(payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_private_snapshot(
    private_snapshot_id: String,
    payload: UpdatePrivateSnapshot,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateSnapshot, String> {
    ensure_private_assets_enabled(&state)?;
    state
        .private_assets_service()
        .update_private_snapshot(&private_snapshot_id, payload)
        .await
        .map_err(|e| e.to_string())
}
